staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go

```diff
@@ -0,0 +1,287 @@
+*/
+
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog/v2"
+)
+
+// TimingRatioHistogram is essentially a gauge for a ratio where the client
+// independently controls the numerator and denominator.
+// When scraped it produces a histogram of samples of the ratio
+// taken at the end of every nanosecond.
+// `*TimingRatioHistogram` implements both Registerable and RatioedGauge.
+type TimingRatioHistogram struct {
+	// The implementation is layered on TimingHistogram,
+	// adding the division by an occasionally adjusted denominator.
+
+	// Registerable is the registerable aspect.
+	// That is the registerable aspect of the underlying TimingHistogram.
+	compbasemetrics.Registerable
+
+	// timingRatioHistogramInner implements the RatioedObsserver aspect.
+	timingRatioHistogramInner
+}
+
+// TimingRatioHistogramOpts is the constructor parameters of a TimingRatioHistogram.
+// The `TimingHistogramOpts.InitialValue` is the initial numerator.
+type TimingRatioHistogramOpts struct {
+	compbasemetrics.TimingHistogramOpts
+	InitialDenominator float64
+}
+
+// timingRatioHistogramInner implements the instrumentation aspect
+type timingRatioHistogramInner struct {
+	nowFunc  func() time.Time
+	delegate interface{ Set(float64) }
+	sync.Mutex
+	numerator, denominator float64 // access only with mutex locked
+}
+
+var _ RatioedGauge = &timingRatioHistogramInner{}
+var _ RatioedGauge = &TimingRatioHistogram{}
+var _ compbasemetrics.Registerable = &TimingRatioHistogram{}
+
+// NewTimingHistogram returns an object which is TimingHistogram-like. However, nothing
+// will be measured until the histogram is registered in at least one registry.
+func NewTimingRatioHistogram(opts *TimingRatioHistogramOpts) *TimingRatioHistogram {
+	return NewTestableTimingRatioHistogram(time.Now, opts)
+}
+
+// NewTestableTimingHistogram adds injection of the clock
+func NewTestableTimingRatioHistogram(nowFunc func() time.Time, opts *TimingRatioHistogramOpts) *TimingRatioHistogram {
+	ratioedOpts := opts.TimingHistogramOpts
+	ratioedOpts.InitialValue /= opts.InitialDenominator
+	th := compbasemetrics.NewTestableTimingHistogram(nowFunc, &ratioedOpts)
+	return &TimingRatioHistogram{
+		Registerable: th,
+		timingRatioHistogramInner: timingRatioHistogramInner{
+			nowFunc:     nowFunc,
+			delegate:    th,
+			numerator:   opts.InitialValue,
+			denominator: opts.InitialDenominator,
+		}}
+}
+
+func (trh *timingRatioHistogramInner) Set(numerator float64) {
+	trh.Lock()
+	defer trh.Unlock()
+	trh.numerator = numerator
+	ratio := numerator / trh.denominator
+	trh.delegate.Set(ratio)
+}
+
+func (trh *timingRatioHistogramInner) Add(deltaNumerator float64) {
+	trh.Lock()
+	defer trh.Unlock()
+	numerator := trh.numerator + deltaNumerator
+	trh.numerator = numerator
+	ratio := numerator / trh.denominator
+	trh.delegate.Set(ratio)
+}
+
+func (trh *timingRatioHistogramInner) Sub(deltaNumerator float64) {
+	trh.Lock()
```

Comment:
Can you just call trh.Add(-deltaNumerator) instead?

Same in Inc/Dec mehods.

Reply:
done in the latest revision